File: backend_php/src/handlers.php
<?php

namespace App;

use Psr\Http\Message\ServerRequestInterface;
use Psr\Http\Message\ResponseInterface;
use Nyholm\Psr7\Response;

class Handlers
{
    private QuoteRepository $repository;

    public function __construct(QuoteRepository $repository)
    {
        $this->repository = $repository;
    }

    private function getClientIp(ServerRequestInterface $request): string
    {
        $headers = $request->getHeader('X-Forwarded-For');
        if (!empty($headers)) {
            $ips = explode(',', $headers[0]);
            return trim($ips[0]);
        }

        $headers = $request->getHeader('X-Real-IP');
        if (!empty($headers)) {
            return $headers[0];
        }

        return '127.0.0.1';
    }

    private function jsonResponse(int $statusCode, array $data): ResponseInterface
    {
        $response = new Response($statusCode);
        $response->getBody()->write(json_encode($data, JSON_UNESCAPED_UNICODE));
        return $response->withHeader('Content-Type', 'application/json')
            ->withHeader('X-Backend', 'php');
    }

    public function getRandom(ServerRequestInterface $request): ResponseInterface
    {
        $quote = $this->repository->getRandom();
        if (!$quote) {
            return $this->jsonResponse(404, ['error' => 'no quotes found']);
        }

        $userIp = $this->getClientIp($request);
        $isLiked = $this->repository->isLiked($quote->id, $userIp);

        return $this->jsonResponse(200, $quote->toResponse($isLiked));
    }

    public function getAll(ServerRequestInterface $request): ResponseInterface
    {
        $queryParams = $request->getQueryParams();
        $page = max(1, (int)($queryParams['page'] ?? 1));
        $pageSize = max(1, min(100, (int)($queryParams['page_size'] ?? 10)));
        $search = $queryParams['search'] ?? null;

        [$quotes, $total] = $this->repository->getAll($page, $pageSize, $search);

        // Оптимизация: batch проверка лайков вместо N+1 запросов
        $userIp = $this->getClientIp($request);
        $quoteIds = array_map(fn($quote) => $quote->id, $quotes);
        $likedMap = $this->repository->areLiked($quoteIds, $userIp);
        
        $responses = [];
        foreach ($quotes as $quote) {
            $isLiked = $likedMap[$quote->id] ?? false;
            $responses[] = $quote->toResponse($isLiked);
        }

        $totalPages = QuoteRepository::calculateTotalPages($total, $pageSize);

        return $this->jsonResponse(200, [
            'quotes' => $responses,
            'total' => $total,
            'page' => $page,
            'page_size' => $pageSize,
            'total_pages' => $totalPages,
        ]);
    }

    public function getById(ServerRequestInterface $request, string $id): ResponseInterface
    {
        $quote = $this->repository->getById($id);
        if (!$quote) {
            return $this->jsonResponse(404, ['error' => 'quote not found']);
        }

        $userIp = $this->getClientIp($request);
        $isLiked = $this->repository->isLiked($quote->id, $userIp);

        return $this->jsonResponse(200, $quote->toResponse($isLiked));
    }

    public function create(ServerRequestInterface $request): ResponseInterface
    {
        $body = json_decode($request->getBody()->getContents(), true);
        if (!isset($body['text']) || !isset($body['author'])) {
            return $this->jsonResponse(400, ['error' => 'text and author are required']);
        }

        // Генерируем UUID в формате, совместимом с другими бэкендами
        $uuid = sprintf(
            '%04x%04x-%04x-%04x-%04x-%04x%04x%04x',
            mt_rand(0, 0xffff), mt_rand(0, 0xffff),
            mt_rand(0, 0xffff),
            mt_rand(0, 0x0fff) | 0x4000,
            mt_rand(0, 0x3fff) | 0x8000,
            mt_rand(0, 0xffff), mt_rand(0, 0xffff), mt_rand(0, 0xffff)
        );

        $quote = new Quote(
            $uuid,
            $body['text'],
            $body['author'],
            0,
            date('c'),
            date('c')
        );

        $this->repository->create($quote);

        return $this->jsonResponse(201, $quote->toResponse(false));
    }

    public function update(ServerRequestInterface $request, string $id): ResponseInterface
    {
        $quote = $this->repository->getById($id);
        if (!$quote) {
            return $this->jsonResponse(404, ['error' => 'quote not found']);
        }

        $body = json_decode($request->getBody()->getContents(), true);
        if (isset($body['text'])) {
            $quote->text = $body['text'];
        }
        if (isset($body['author'])) {
            $quote->author = $body['author'];
        }
        $quote->updatedAt = date('c');

        if (!$this->repository->update($id, $quote)) {
            return $this->jsonResponse(500, ['error' => 'internal server error']);
        }

        $updatedQuote = $this->repository->getById($id);
        $userIp = $this->getClientIp($request);
        $isLiked = $this->repository->isLiked($updatedQuote->id, $userIp);

        return $this->jsonResponse(200, $updatedQuote->toResponse($isLiked));
    }

    public function delete(ServerRequestInterface $request, string $id): ResponseInterface
    {
        if (!$this->repository->delete($id)) {
            return $this->jsonResponse(404, ['error' => 'quote not found']);
        }

        return new Response(204);
    }

    public function like(ServerRequestInterface $request, string $id): ResponseInterface
    {
        $userIp = $this->getClientIp($request);
        $userAgent = $request->getHeader('User-Agent')[0] ?? null;

        try {
            $this->repository->like($id, $userIp, $userAgent);
        } catch (\RuntimeException $e) {
            if (strpos($e->getMessage(), 'already liked') !== false) {
                return $this->jsonResponse(400, ['error' => 'Вы уже поставили лайк этой цитате']);
            }
            if (strpos($e->getMessage(), 'not found') !== false) {
                return $this->jsonResponse(404, ['error' => 'quote not found']);
            }
            return $this->jsonResponse(500, ['error' => 'internal server error']);
        }

        $quote = $this->repository->getById($id);
        return $this->jsonResponse(200, $quote->toResponse(true));
    }

    public function getTopWeekly(ServerRequestInterface $request): ResponseInterface
    {
        $quote = $this->repository->getTopWeekly();
        if (!$quote) {
            return $this->jsonResponse(404, ['error' => 'no quotes found for the last week']);
        }

        $userIp = $this->getClientIp($request);
        $isLiked = $this->repository->isLiked($quote->id, $userIp);

        return $this->jsonResponse(200, $quote->toResponse($isLiked));
    }

    public function getTopAllTime(ServerRequestInterface $request): ResponseInterface
    {
        $quote = $this->repository->getTopAllTime();
        if (!$quote) {
            return $this->jsonResponse(404, ['error' => 'no quotes found']);
        }

        $userIp = $this->getClientIp($request);
        $isLiked = $this->repository->isLiked($quote->id, $userIp);

        return $this->jsonResponse(200, $quote->toResponse($isLiked));
    }

    public function resetLikes(ServerRequestInterface $request): ResponseInterface
    {
        $this->repository->resetLikes();
        return $this->jsonResponse(200, ['message' => 'Все лайки успешно сброшены']);
    }

    public function health(ServerRequestInterface $request): ResponseInterface
    {
        return $this->jsonResponse(200, ['status' => 'ok']);
    }
}

File: backend_cpp/src/handlers.cpp
#include "handlers.h"
#include <drogon/HttpResponse.h>
#include <drogon/HttpTypes.h>
#include <drogon/utils/Utilities.h>
#include <json/json.h>
#include <uuid/uuid.h>
#include <ctime>
#include <iomanip>
#include <sstream>
#include <random>

std::string QuoteHandlers::getClientIP(const drogon::HttpRequestPtr& req) {
    auto forwarded = req->getHeader("X-Forwarded-For");
    if (!forwarded.empty()) {
        size_t pos = forwarded.find(',');
        return pos != std::string::npos ? forwarded.substr(0, pos) : forwarded;
    }
    
    auto realIP = req->getHeader("X-Real-IP");
    if (!realIP.empty()) {
        return realIP;
    }
    
    return "127.0.0.1";
}

void QuoteHandlers::getRandom(const drogon::HttpRequestPtr& req,
                              std::function<void(const drogon::HttpResponsePtr&)>&& callback) {
    try {
        Quote quote = repo_->getRandom();
        std::string userIP = getClientIP(req);
        bool isLiked = repo_->isLiked(quote.id, userIP);
        
        Json::Value json;
        json["id"] = quote.id;
        json["text"] = quote.text;
        json["author"] = quote.author;
        json["likes_count"] = quote.likesCount;
        json["is_liked"] = isLiked;
        json["created_at"] = quote.createdAt;
        json["updated_at"] = quote.updatedAt;
        
        auto resp = drogon::HttpResponse::newHttpJsonResponse(json);
        resp->addHeader("X-Backend", "cpp");
        callback(resp);
    } catch (const std::exception& e) {
        Json::Value error;
        error["error"] = e.what();
        auto resp = drogon::HttpResponse::newHttpJsonResponse(error);
        resp->setStatusCode(drogon::k404NotFound);
        resp->addHeader("X-Backend", "cpp");
        callback(resp);
    }
}

void QuoteHandlers::getAll(const drogon::HttpRequestPtr& req,
                           std::function<void(const drogon::HttpResponsePtr&)>&& callback) {
    try {
        auto& params = req->getParameters();
        int page = 1;
        int pageSize = 10;
        std::string search;
        
        auto pageIt = params.find("page");
        if (pageIt != params.end()) {
            page = std::max(1, std::stoi(pageIt->second));
        }
        auto pageSizeIt = params.find("page_size");
        if (pageSizeIt != params.end()) {
            pageSize = std::max(1, std::min(100, std::stoi(pageSizeIt->second)));
        }
        auto searchIt = params.find("search");
        if (searchIt != params.end()) {
            search = searchIt->second;
        }
        
        auto [quotes, total] = repo_->getAll(page, pageSize, search);
        std::string userIP = getClientIP(req);
        
        // Batch check likes
        std::vector<std::string> quoteIds;
        quoteIds.reserve(quotes.size());
        for (const auto& quote : quotes) {
            quoteIds.push_back(quote.id);
        }
        auto likedMap = repo_->areLiked(quoteIds, userIP);
        
        Json::Value json;
        Json::Value quotesArray(Json::arrayValue);
        for (const auto& quote : quotes) {
            Json::Value quoteJson;
            quoteJson["id"] = quote.id;
            quoteJson["text"] = quote.text;
            quoteJson["author"] = quote.author;
            quoteJson["likes_count"] = quote.likesCount;
            quoteJson["is_liked"] = likedMap[quote.id];
            quoteJson["created_at"] = quote.createdAt;
            quoteJson["updated_at"] = quote.updatedAt;
            quotesArray.append(quoteJson);
        }
        
        json["quotes"] = quotesArray;
        json["total"] = static_cast<Json::Int64>(total);
        json["page"] = page;
        json["page_size"] = pageSize;
        json["total_pages"] = static_cast<int>(calculateTotalPages(total, pageSize));
        
        auto resp = drogon::HttpResponse::newHttpJsonResponse(json);
        resp->addHeader("X-Backend", "cpp");
        callback(resp);
    } catch (const std::exception& e) {
        Json::Value error;
        error["error"] = e.what();
        auto resp = drogon::HttpResponse::newHttpJsonResponse(error);
        resp->setStatusCode(drogon::k500InternalServerError);
        resp->addHeader("X-Backend", "cpp");
        callback(resp);
    }
}

void QuoteHandlers::getById(const drogon::HttpRequestPtr& req,
                            std::function<void(const drogon::HttpResponsePtr&)>&& callback,
                            const std::string& id) {
    try {
        Quote quote = repo_->getById(id);
        std::string userIP = getClientIP(req);
        bool isLiked = repo_->isLiked(quote.id, userIP);
        
        Json::Value json;
        json["id"] = quote.id;
        json["text"] = quote.text;
        json["author"] = quote.author;
        json["likes_count"] = quote.likesCount;
        json["is_liked"] = isLiked;
        json["created_at"] = quote.createdAt;
        json["updated_at"] = quote.updatedAt;
        
        auto resp = drogon::HttpResponse::newHttpJsonResponse(json);
        resp->addHeader("X-Backend", "cpp");
        callback(resp);
    } catch (const std::exception& e) {
        Json::Value error;
        error["error"] = e.what();
        auto resp = drogon::HttpResponse::newHttpJsonResponse(error);
        resp->setStatusCode(drogon::k404NotFound);
        resp->addHeader("X-Backend", "cpp");
        callback(resp);
    }
}

void QuoteHandlers::create(const drogon::HttpRequestPtr& req,
                           std::function<void(const drogon::HttpResponsePtr&)>&& callback) {
    try {
        auto jsonPtr = req->getJsonObject();
        if (!jsonPtr || !jsonPtr->isMember("text") || !jsonPtr->isMember("author")) {
            Json::Value error;
            error["error"] = "text and author are required";
            auto resp = drogon::HttpResponse::newHttpJsonResponse(error);
            resp->setStatusCode(drogon::k400BadRequest);
            resp->addHeader("X-Backend", "cpp");
            callback(resp);
            return;
        }
        
        // Generate UUID v4
        std::random_device rd;
        std::mt19937 gen(rd());
        std::uniform_int_distribution<> dis(0, 15);
        std::uniform_int_distribution<> dis2(8, 11);
        
        std::stringstream uuidStream;
        for (int i = 0; i < 32; ++i) {
            if (i == 8 || i == 12 || i == 16 || i == 20) {
                uuidStream << "-";
            }
            int v = (i == 12) ? dis2(gen) : dis(gen);
            uuidStream << std::hex << v;
        }
        std::string uuidStr = uuidStream.str();
        
        // Get current time
        auto now = std::time(nullptr);
        std::stringstream ss;
        ss << std::put_time(std::gmtime(&now), "%Y-%m-%d %H:%M:%S");
        std::string timestamp = ss.str();
        
        Quote quote;
        quote.id = uuidStr;
        quote.text = (*jsonPtr)["text"].asString();
        quote.author = (*jsonPtr)["author"].asString();
        quote.likesCount = 0;
        quote.createdAt = timestamp;
        quote.updatedAt = timestamp;
        
        repo_->create(quote);
        
        Json::Value json;
        json["id"] = quote.id;
        json["text"] = quote.text;
        json["author"] = quote.author;
        json["likes_count"] = quote.likesCount;
        json["is_liked"] = false;
        json["created_at"] = quote.createdAt;
        json["updated_at"] = quote.updatedAt;
        
        auto resp = drogon::HttpResponse::newHttpJsonResponse(json);
        resp->setStatusCode(drogon::k201Created);
        resp->addHeader("X-Backend", "cpp");
        callback(resp);
    } catch (const std::exception& e) {
        Json::Value error;
        error["error"] = e.what();
        auto resp = drogon::HttpResponse::newHttpJsonResponse(error);
        resp->setStatusCode(drogon::k500InternalServerError);
        resp->addHeader("X-Backend", "cpp");
        callback(resp);
    }
}

void QuoteHandlers::update(const drogon::HttpRequestPtr& req,
                           std::function<void(const drogon::HttpResponsePtr&)>&& callback,
                           const std::string& id) {
    try {
        Quote quote = repo_->getById(id);
        auto jsonPtr = req->getJsonObject();
        
        if (jsonPtr && jsonPtr->isMember("text")) {
            quote.text = (*jsonPtr)["text"].asString();
        }
        if (jsonPtr && jsonPtr->isMember("author")) {
            quote.author = (*jsonPtr)["author"].asString();
        }
        
        auto now = std::time(nullptr);
        std::stringstream ss;
        ss << std::put_time(std::gmtime(&now), "%Y-%m-%d %H:%M:%S");
        quote.updatedAt = ss.str();
        
        repo_->update(id, quote);
        
        Quote updatedQuote = repo_->getById(id);
        std::string userIP = getClientIP(req);
        bool isLiked = repo_->isLiked(updatedQuote.id, userIP);
        
        Json::Value json;
        json["id"] = updatedQuote.id;
        json["text"] = updatedQuote.text;
        json["author"] = updatedQuote.author;
        json["likes_count"] = updatedQuote.likesCount;
        json["is_liked"] = isLiked;
        json["created_at"] = updatedQuote.createdAt;
        json["updated_at"] = updatedQuote.updatedAt;
        
        auto resp = drogon::HttpResponse::newHttpJsonResponse(json);
        resp->addHeader("X-Backend", "cpp");
        callback(resp);
    } catch (const std::exception& e) {
        Json::Value error;
        error["error"] = e.what();
        auto resp = drogon::HttpResponse::newHttpJsonResponse(error);
        resp->setStatusCode(drogon::k404NotFound);
        resp->addHeader("X-Backend", "cpp");
        callback(resp);
    }
}

void QuoteHandlers::deleteQuote(const drogon::HttpRequestPtr& req,
                                std::function<void(const drogon::HttpResponsePtr&)>&& callback,
                                const std::string& id) {
    try {
        repo_->deleteQuote(id);
        auto resp = drogon::HttpResponse::newHttpResponse();
        resp->setStatusCode(drogon::k204NoContent);
        resp->addHeader("X-Backend", "cpp");
        callback(resp);
    } catch (const std::exception& e) {
        Json::Value error;
        error["error"] = e.what();
        auto resp = drogon::HttpResponse::newHttpJsonResponse(error);
        resp->setStatusCode(drogon::k404NotFound);
        resp->addHeader("X-Backend", "cpp");
        callback(resp);
    }
}

void QuoteHandlers::like(const drogon::HttpRequestPtr& req,
                         std::function<void(const drogon::HttpResponsePtr&)>&& callback,
                         const std::string& id) {
    try {
        std::string userIP = getClientIP(req);
        std::string userAgent = req->getHeader("User-Agent");
        
        repo_->like(id, userIP, userAgent);
        
        Quote quote = repo_->getById(id);
        
        Json::Value json;
        json["id"] = quote.id;
        json["text"] = quote.text;
        json["author"] = quote.author;
        json["likes_count"] = quote.likesCount;
        json["is_liked"] = true;
        json["created_at"] = quote.createdAt;
        json["updated_at"] = quote.updatedAt;
        
        auto resp = drogon::HttpResponse::newHttpJsonResponse(json);
        resp->addHeader("X-Backend", "cpp");
        callback(resp);
    } catch (const std::exception& e) {
        std::string errorMsg = e.what();
        Json::Value error;
        
        if (errorMsg.find("already liked") != std::string::npos) {
            error["error"] = "Вы уже поставили лайк этой цитате";
            auto resp = drogon::HttpResponse::newHttpJsonResponse(error);
            resp->setStatusCode(drogon::k400BadRequest);
            resp->addHeader("X-Backend", "cpp");
            callback(resp);
        } else if (errorMsg.find("not found") != std::string::npos) {
            error["error"] = "quote not found";
            auto resp = drogon::HttpResponse::newHttpJsonResponse(error);
            resp->setStatusCode(drogon::k404NotFound);
            resp->addHeader("X-Backend", "cpp");
            callback(resp);
        } else {
            error["error"] = "internal server error";
            auto resp = drogon::HttpResponse::newHttpJsonResponse(error);
            resp->setStatusCode(drogon::k500InternalServerError);
            resp->addHeader("X-Backend", "cpp");
            callback(resp);
        }
    }
}

void QuoteHandlers::getTopWeekly(const drogon::HttpRequestPtr& req,
                                 std::function<void(const drogon::HttpResponsePtr&)>&& callback) {
    try {
        Quote quote = repo_->getTopWeekly();
        std::string userIP = getClientIP(req);
        bool isLiked = repo_->isLiked(quote.id, userIP);
        
        Json::Value json;
        json["id"] = quote.id;
        json["text"] = quote.text;
        json["author"] = quote.author;
        json["likes_count"] = quote.likesCount;
        json["is_liked"] = isLiked;
        json["created_at"] = quote.createdAt;
        json["updated_at"] = quote.updatedAt;
        
        auto resp = drogon::HttpResponse::newHttpJsonResponse(json);
        resp->addHeader("X-Backend", "cpp");
        callback(resp);
    } catch (const std::exception& e) {
        Json::Value error;
        error["error"] = e.what();
        auto resp = drogon::HttpResponse::newHttpJsonResponse(error);
        resp->setStatusCode(drogon::k404NotFound);
        resp->addHeader("X-Backend", "cpp");
        callback(resp);
    }
}

void QuoteHandlers::getTopAllTime(const drogon::HttpRequestPtr& req,
                                  std::function<void(const drogon::HttpResponsePtr&)>&& callback) {
    try {
        Quote quote = repo_->getTopAllTime();
        std::string userIP = getClientIP(req);
        bool isLiked = repo_->isLiked(quote.id, userIP);
        
        Json::Value json;
        json["id"] = quote.id;
        json["text"] = quote.text;
        json["author"] = quote.author;
        json["likes_count"] = quote.likesCount;
        json["is_liked"] = isLiked;
        json["created_at"] = quote.createdAt;
        json["updated_at"] = quote.updatedAt;
        
        auto resp = drogon::HttpResponse::newHttpJsonResponse(json);
        resp->addHeader("X-Backend", "cpp");
        callback(resp);
    } catch (const std::exception& e) {
        Json::Value error;
        error["error"] = e.what();
        auto resp = drogon::HttpResponse::newHttpJsonResponse(error);
        resp->setStatusCode(drogon::k404NotFound);
        resp->addHeader("X-Backend", "cpp");
        callback(resp);
    }
}

void QuoteHandlers::resetLikes(const drogon::HttpRequestPtr& req,
                               std::function<void(const drogon::HttpResponsePtr&)>&& callback) {
    try {
        repo_->resetLikes();
        Json::Value json;
        json["message"] = "Все лайки успешно сброшены";
        auto resp = drogon::HttpResponse::newHttpJsonResponse(json);
        resp->addHeader("X-Backend", "cpp");
        callback(resp);
    } catch (const std::exception& e) {
        Json::Value error;
        error["error"] = e.what();
        auto resp = drogon::HttpResponse::newHttpJsonResponse(error);
        resp->setStatusCode(drogon::k500InternalServerError);
        resp->addHeader("X-Backend", "cpp");
        callback(resp);
    }
}

void QuoteHandlers::health(const drogon::HttpRequestPtr& req,
                           std::function<void(const drogon::HttpResponsePtr&)>&& callback) {
    Json::Value json;
    json["status"] = "ok";
    auto resp = drogon::HttpResponse::newHttpJsonResponse(json);
    resp->addHeader("X-Backend", "cpp");
    callback(resp);
}

